server/src/logic/transaction_operations.rs

```diff
@@ -0,0 +1,97 @@
+use super::common;
+use super::error::ExpenseTransactionError;
+use super::user_operations::authorize_user_by_id;
+use crate::routes::dto::transactions::NewTransactionRequest;
+
+use entity::currency_type::{self, Entity as CurrencyType};
+use entity::expense::{self, Entity as Expense};
+use entity::transaction::{self, Entity as Transaction};
+use entity::Id;
+
+use chrono::NaiveDate;
+use sea_orm::ActiveValue::NotSet;
+use sea_orm::{ActiveModelTrait, ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter, Set};
+
+pub async fn create_transaction(
+    conn: &DatabaseConnection,
+    user_id: Id,
+    req: NewTransactionRequest,
+) -> Result<Id, ExpenseTransactionError> {
+    let (referred_expense, currency_validation_result) = tokio::join!(
+        get_expense_by_id_if_exists(conn, req.expense_id),
+        validate_referred_currency_type(conn, req.currency_type_id)
+    );
+
+    currency_validation_result?;
+    authorize_user_by_id(user_id, referred_expense?.user_id)?;
```

Comment:
This looks very idiomatic, nice job :smile:

Reply:
😎